### ISSUE TITLE
[Question] Is it possible to override the models' data loading methods?

### ISSUE THREAD

**Issue by giacomoguiduzzi** [unit8co/darts#2365]
Hi, I already tried searching on the web about this but I couldn't get a clear idea of what's going on and how to proceed. I'm trying to create a small benchmarking software that takes some _Darts_ models and datasets, prepares the data and feeds it to the models so to gather some metrics. I read that through a custom `PyTorch Lightning Trainer` it is possible to pass custom `DataLoaders`, so my idea was to use a custom `Trainer` to use the data that I prepared with sliding windows. I essentially wanted to override the models' default data loading with my `DataLoader` through the trainer. However, I noticed that the custom `Trainer` requires a `PyTorchLightningModule` class, while my model is (for example) a `NBEATSModel` class.  I assume that the sliding window processing is done under the hood by the default DataLoaders _Darts_ provides to models, but I'd like to prepare my own sliding windows as they may have missing values and I want to handle them myself.
To sum up: is there some way I can customize how data is loaded in the models, passing my own sliding window dataset and my own DataLoader?
Thank you in advance, I'd really appreciate any help you can provide.

Giacomo Guiduzzi

**Comment by dennisbader** [comment]
Hi @giacomoguiduzzi, you could write custom training and inference datasets and then pass them to [`fit_from_dataset`](https://unit8co.github.io/darts/generated_api/darts.models.forecasting.nbeats.html#darts.models.forecasting.nbeats.NBEATSModel.fit_from_dataset), and [`predict_from_dataset()`](https://unit8co.github.io/darts/generated_api/darts.models.forecasting.nbeats.html#darts.models.forecasting.nbeats.NBEATSModel.predict_from_dataset). All training datasets make use of [`GenericShiftedDataset`](https://unit8co.github.io/darts/_modules/darts/utils/data/shifted_dataset.html#GenericShiftedDataset), and [`GenericInferenceDataset`](https://unit8co.github.io/darts/_modules/darts/utils/data/inference_dataset.html#GenericInferenceDataset). You can use those as a reference. For `NBEATSModel`, the final datasets are [`PastCovariatesSequentialDataset`](https://unit8co.github.io/darts/_modules/darts/utils/data/sequential_dataset.html#PastCovariatesSequentialDataset), and [`PastCovariatesInferenceDataset `](https://unit8co.github.io/darts/_modules/darts/utils/data/inference_dataset.html#PastCovariatesInferenceDataset).

It's currently not possible to use a custom DataLoader with these methods, but you should be able to implement the desired logic (sliding windows, skipping missing value handling, ...) through the dataset itself.

**Comment by igorrivin** [comment]
Well, on this subject, when I run the NHiTSModel, Lightning tells me that I should increase "num_workers" in the dataloader, but for the life of me I cannot figure out how to do it through darts. pl_trainer_args are not it, it seems.

**Comment by madtoinou** [comment]
@igorrivin Someone opened a PR to add this feature (#2295), we are trying to free some time to review it. Note that despite this hint from Lightning, it does not always improve performance.

@giacomoguiduzzi To add on @dennisbader's answer; the `DataLoader` is not responsible for converting the `TimeSeries` into chunks but just to serve them, as batch, shuffled, ... It's the `[...]Dataset` class that is responsible for generating these chunks and shaping its dimensions (often the covariates) into something suitable for the model ([pytorch doc](https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset)). Make sure that your changes are reflected in both the `Training` and `Inference` dataset classes.

**Comment by giacomoguiduzzi** [comment]
Thank you all for the info @dennisbader @igorrivin @madtoinou! I'll try changing my implementation and exploiting those Dataset classes. Also, thanks @madtoinou for the clarification, I like my code to be clean and being new to PyTorch is making it pretty difficult to achieve. 
Have a great day!
